Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -5,182 +5,277 @@
--- the Quadratic Sieve attempts to output @factor@, a factor of @n@. The
+data QuadraticSieveConfig = QuadraticSieveConfig
+  { qscFactorBase :: Int
+  , qscSievingInterval :: Int
+  , qscNumberOfBlocks :: Int
+  , qscSmoothThreshold :: Int
+  }
+
+autoConfig :: Integer -> QuadraticSieveConfig
+autoConfig n = QuadraticSieveConfig t m k h
+  where
+    -- + 4 for large prime variation
+    h = intLog2 t + 1
+    k = max 1 (l `div` 10)
+    m = 3 * t `div` 2
+    t
+      | l < 4    = integerToInt n `div` 2
+      | l < 8    = integerToInt $ integerSquareRoot n
+      | otherwise = (50 - l) * floor (sqrt . exp . sqrt $ log (fromInteger n) * log (log (fromInteger n)) :: Double)
+    -- number of digits of n
+    l = integerLog10 n
+
+-- | Given an odd positive composite Integer @n@ and Int parameters @b@, @t@ and
+-- @k@, the Quadratic Sieve attempts to output @factor@, a factor of @n@. The
 -- parameter @b@ controls the size of the factor base. This consists of all
 -- the relevant primes which are less than or equal to @b@. The parameter @t@
--- controls the length of the sieving interval.
-quadraticSieve :: Integer -> Int -> Int -> Integer
-quadraticSieve n b t = findFactor n $ findSquares n b t
+-- controls the length of the sieving interval. The parameter @k@ controls the
+-- maximum number of blocks for the sieve to go through (@2 ^ (k - 1)@).
+quadraticSieve :: Integer -> Integer
+quadraticSieve n = quadraticSieveManual n $ autoConfig n
+
+quadraticSieveManual :: Integer -> QuadraticSieveConfig -> Integer
+quadraticSieveManual n qsc = findFactor n $ findSquares n qsc
 
 findFactor :: Integer -> [(Integer, Integer)] -> Integer
-findFactor _ [] = error "Quadratic Sieve failed."
+findFactor _ [] = error "Parameters are not large enough."
 findFactor n ((x, y) : otherSquares)
-  | factor /= 1 && factor /= n = factor
-  | otherwise                  = findFactor n otherSquares
+  | (x * x - y * y) `mod` n /= 0 = error ("Algorithm incorrect." ++ show (x, y))
+  | factor /= 1 && factor /= n   = factor
+  | otherwise                    = findFactor n otherSquares
   where
     factor = gcd (x - y) n
 
 -- This routine outputs an infinite list of tuples @(x, y)@ such that
 -- @x ^ 2 - y ^ 2 `mod` n = 0@. A factorisation can be infered from this data
 -- in at least a half of the cases.
-findSquares :: Integer -> Int -> Int -> [(Integer, Integer)]
-findSquares n b t = runST $ do
+findSquares :: Integer -> QuadraticSieveConfig -> [(Integer, Integer)]
+-- k indicates maximum number of sieve blocks to go thorugh 2 ^ (k - 1)
+findSquares n (QuadraticSieveConfig t m k h) = runST $ do
   let
-    factorBase = [nextPrime 2..precPrime b]
-    squareRoot = integerSquareRoot n
-    mappingFunction j = j * j - n
-    startingPoint = squareRoot - intToInteger t `div` 2
+    factorBase = filter (\p -> unPrime p == 2 || jacobi n ((intToInteger . unPrime) p) == One) [nextPrime 2..precPrime t]
```

Comment:
Let's define `isResidue :: Integer -> Prime Int -> Bool` as a separate helper function and use it here and below.